File: dbgear/models/schema.py
from .base import BaseSchema


class Field(BaseSchema):
    column_name: str
    display_name: str
    column_type: str
    nullable: bool
    primary_key: int | None
    default_value: str | None
    foreign_key: str | None
    comment: str | None


class Index(BaseSchema):
    index_name: str | None
    columns: list[str]


class Table(BaseSchema):
    instance: str
    table_name: str
    display_name: str
    fields: list[Field] = []
    indexes: list[Index] = []
    # FIXME 参照元をデータとして持たせるか？


class Schema:

    def __init__(self, name):
        self.name = name
        self.tables = {}

    def __repr__(self) -> str:
        return f'{self.tables}'

    def add_table(self, table: Table) -> None:
        self.tables[table.table_name] = table

    def get_table(self, name: str) -> Table:
        return self.tables[name]

    def get_tables(self) -> dict[str, Table]:
        return self.tables


def find_field(fields: list[Field], name: str):
    field = next(filter(lambda x: x.column_name == name, fields), None)
    if field is None:
        raise RuntimeError(f'Could not find field. ({name})')
    return field

File: packages/frontend/src/types/schema.ts
export interface Field {
  column_name: string
  display_name: string | null
  column_type: string
  nullable: boolean
  primary_key: number | null
  default_value: string | null
  foreign_key: string | null
  comment: string | null
  expression: string | null
  stored: boolean | null
  auto_increment: boolean | null
  charset: string | null
  collation: string | null
}

export interface Index {
  index_name: string
  columns: string[]
}

export interface Table {
  instance: string
  table_name: string
  display_name: string | null
  fields: Field[]
  indexes: Index[]
}

export interface Schema {
  name: string
  tables: Record<string, Table>
  views: Record<string, any>
}

export interface TableFormData {
  table_name: string
  display_name: string
  fields: Field[]
  indexes: Index[]
}

export interface FieldFormData {
  column_name: string
  display_name: string
  column_type: string
  nullable: boolean
  primary_key: string
  auto_increment: boolean
  default_value: string
  foreign_key_table: string
  foreign_key_column: string
  comment: string
  expression: string
  stored: string
  charset: string
  collation: string
}